openghg/storage/_indexing.py

```diff
@@ -63,3 +63,174 @@ def select_nonoverlaps(self, ds: xr.Dataset, dim: str) -> xr.Dataset:
             raise ValueError(f"Dimension {dim} not found.") from e
 
         return ds.where(ds[dim][self.nonoverlaps(other)], drop=True)
+
+
+# ----------------------------------------
+# Finding contiguous regions
+# ----------------------------------------
+
+
+def _sort_by_target(source: np.ndarray, target: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+    """Sort `target` array and apply the same permutation to `source`."""
+    shuf = np.argsort(target)
+    return source[shuf], target[shuf]
+
+
+def is_monotonic(idx: pd.Index) -> bool:
+    """Return True if index is increasing or decreasing."""
+    # explicitly convert to bool for mypy...
+    return bool(idx.is_monotonic_increasing or idx.is_monotonic_decreasing)
+
+
+class IndexingError(Exception):
+    """Error raised by indexing methods."""
+
+    pass
+
+
+def _alignment_indexers(
+    source: pd.Index | np.ndarray,
+    target: pd.Index,
+    ignore_missing: bool = False,
+    **kwargs: Any,
+) -> tuple[np.ndarray, np.ndarray]:
+    """Get indexers for aligning source and target."""
+    if kwargs and not is_monotonic(target):
+        raise ValueError("Cannot align to unsorted index with tolerances.")
+
+    # type hints to help mypy... this shouldn't be necessary with numpy >= 2.3
+    # and mypy >= 1.16.
+    # see: https://github.com/numpy/numpy/issues/28076
+    source_idxer: np.typing.NDArray[np.integer] = np.arange(len(source))
+    align_idxer: np.typing.NDArray[np.integer] = target.get_indexer(source, **kwargs)
```

Comment:
Can you add a comment / detail in the docstring that `-1` being returned means the index hasn't been found? Seems like a useful detail which could get lost if we don't record it now.